src/celeritas/ext/ScopedGeantLogger.cc

```diff
@@ -0,0 +1,159 @@
+    return 0;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Send Geant4 log messages to Celeritas' world logger.
+ */
+class GeantLoggerAdapter : public G4coutDestination
+{
+  public:
+    // Assign to Geant handlers on construction
+    GeantLoggerAdapter();
+    ~GeantLoggerAdapter();
+
+    // Handle error messages
+    G4int ReceiveG4cout(G4String const& str) final;
+    G4int ReceiveG4cerr(G4String const& str) final;
+
+  private:
+#if CELER_G4SSBUF
+    G4coutDestination* saved_cout_{nullptr};
+    G4coutDestination* saved_cerr_{nullptr};
+#endif
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Redirect geant4's stdout/cerr on construction.
+ */
+GeantLoggerAdapter::GeantLoggerAdapter()
+{
+#if CELER_G4SSBUF
+    saved_cout_ = G4coutbuf.GetDestination();
+    saved_cerr_ = G4cerrbuf.GetDestination();
+    G4coutbuf.SetDestination(this);
+    G4cerrbuf.SetDestination(this);
+#else
+    // See Geant4 change global-V11-01-01
+    G4iosSetDestination(this);
+#endif
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Restore iostream buffers on destruction.
+ */
+GeantLoggerAdapter::~GeantLoggerAdapter()
+{
+#if CELER_G4SSBUF
+    G4coutbuf.SetDestination(saved_cout_);
+    G4cerrbuf.SetDestination(saved_cerr_);
+#else
+    G4iosSetDestination(nullptr);
+#endif
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Process stdout message.
+ */
+G4int GeantLoggerAdapter::ReceiveG4cout(G4String const& str)
+{
+    return log_impl(str, LogLevel::diagnostic);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Process stderr message.
+ */
+G4int GeantLoggerAdapter::ReceiveG4cerr(G4String const& str)
+{
+    return log_impl(str, LogLevel::info);
+}
+
+//---------------------------------------------------------------------------//
+//! Global flag for "ownership" of the Geant4 logger
+bool g_adapter_active_{false};
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Install the Celeritas Geant4 logger.
+ *
+ * A global flag allows multiple logger adapters to be nested without
+ * consequence.
+ */
+ScopedGeantLogger::ScopedGeantLogger()
+{
+    if (!g_adapter_active_)
+    {
+        static std::mutex capture_mutex;
+        std::lock_guard<std::mutex> scoped_lock{capture_mutex};
+
+        if (!g_adapter_active_)
+        {
+            g_adapter_active_ = true;
+            logger_ = std::make_unique<GeantLoggerAdapter>();
+        }
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Revert to the previous exception handler.
+ */
+ScopedGeantLogger::~ScopedGeantLogger()
+{
+    if (logger_)
+    {
+        logger_.reset();
```

Comment:
Do we need to take the `capture_mutex` lock here?

Reply:
Not here but below.  Technically `g_adapter_active_` needs to either be an atomic or its update ought to be covered by the mutex.  In practice it may have no important visible effect (i.e. at worse another thread may not be able to install a the logger even-though it could of but it would also be the case with small variation in the timing of the threads).

Reply:
I realized after a little more digging that the G4 output elements are thread-local: so we don't need mutexes at all, but the `g_adapter_active_` needs to be thread local. I've adjusted accordingly.

Reply:
Which commit implement this adjustment?

Reply:
Uhhhh apparently one that I didn't actually `git commit` 😞 this is why we have reviews... thanks for catching that!